File: src/MixNetFx/ADSqlSupportComponents/AdDatabaseManager.cs
using System;
using System.Data.SqlClient;

namespace ActiveDev.Data.SqlClient
{
    public class ADSqlDatabaseManager
    {
        private readonly string _sqlInstanceConnString;
        private readonly string _databaseName;
        private string _filenameOnSqlServer;
        private bool _databaseExists;
        private bool _lastSqlResult;
        private Exception _lastSqlException;

        public ADSqlDatabaseManager(string sqlInstanceConnString, string databasename)
        {
            _sqlInstanceConnString = sqlInstanceConnString;
            _databaseName = databasename;
        }

        public void QueryProperties()
        {
            var connection = new SqlConnection(_sqlInstanceConnString);
            SqlCommand command;

            using (connection)
            {
                try
                {
                    connection.Open();
                    _lastSqlResult = true;
                }
                catch (Exception ex)
                {
                    _lastSqlException = ex;
                    _lastSqlResult = false;
                    return;
                }

                command = new SqlCommand("SELECT name FROM sys.databases where name = @name");
                command.Parameters.Add("@name", System.Data.SqlDbType.NVarChar, 128).Value = _databaseName;
                command.Connection = connection;
                SqlDataReader reader = command.ExecuteReader();
                _databaseExists = reader.HasRows;
                reader.Close();
            }

            if (_databaseExists)
            {
                connection = new SqlConnection(_sqlInstanceConnString + "; Initial Catalog='" + _databaseName + "'");
                using (connection)
                {
                    try
                    {
                        connection.Open();
                        _lastSqlResult = true;
                    }
                    catch (Exception ex)
                    {
                        _lastSqlException = ex;
                        _lastSqlResult = false;
                        return;
                    }

                    command = new SqlCommand("SELECT name, physical_name FROM sys.database_files where name = @name");
                    command.Parameters.Add("@name", System.Data.SqlDbType.NVarChar, 128).Value = _databaseName;
                    command.Connection = connection;
                    SqlDataReader reader = command.ExecuteReader();
                    if (reader.HasRows)
                    {
                        reader.Read();
                        _filenameOnSqlServer = reader.GetString(reader.GetOrdinal("physical_name"));
                    }
                    reader.Close();
                }
            }
            _lastSqlResult = true;
        }

        public static ADSqlDatabaseManager AttachDatabase(string sqlInstanceConnString, string databasename, string databaseFilename)
        {
            return AttachDatabase(sqlInstanceConnString, databasename, databaseFilename, null, null);
        }

        public static ADSqlDatabaseManager AttachDatabase(string sqlInstanceConnString, string databasename, string databaseFilename, string newDbOwner)
        {
            return AttachDatabase(sqlInstanceConnString, databasename, databaseFilename, null, newDbOwner);
        }

        public static ADSqlDatabaseManager AttachDatabase(string sqlInstanceConnString, string databasename,
            string databaseFilename, string logFilename, string newDbOwner)
        {
            var connection = new SqlConnection(sqlInstanceConnString);
            SqlCommand command;

            if (string.IsNullOrEmpty(logFilename))
            {
                command = new SqlCommand(
                    BuildAttachDatabaseSql(databasename, databaseFilename),
                    connection);
            }
            else
            {
                command = new SqlCommand(
                    BuildAttachDatabaseWithLogSql(databasename, databaseFilename, logFilename),
                    connection);
            }

            using (connection)
            {
                connection.Open();
                command.ExecuteScalar();

                if (!string.IsNullOrEmpty(newDbOwner))
                {
                    command = new SqlCommand(
                        BuildChangeDbOwnerSql(databasename, newDbOwner),
                        connection);
                    command.ExecuteScalar();
                }
            }
            return new ADSqlDatabaseManager(sqlInstanceConnString, databasename);
        }

        private static string BuildAttachDatabaseSql(string databaseName, string databaseFilename)
        {
            return "CREATE DATABASE " + QuoteIdentifier(databaseName) + " ON " +
                   "( FILENAME = N'" + EscapeStringLiteral(databaseFilename) + "' )" +
                   " FOR ATTACH_REBUILD_LOG";
        }

        private static string BuildAttachDatabaseWithLogSql(string databaseName, string databaseFilename, string logFilename)
        {
            return "CREATE DATABASE " + QuoteIdentifier(databaseName) + " ON " +
                   "( FILENAME = N'" + EscapeStringLiteral(databaseFilename) + "' )," +
                   "( FILENAME = N'" + EscapeStringLiteral(logFilename) + "' )" +
                   " FOR ATTACH";
        }

        private static string BuildChangeDbOwnerSql(string databaseName, string newDbOwner)
        {
            return "if not exists (select name from master.sys.databases sd where name = @dbname and SUSER_SNAME(sd.owner_sid) = SUSER_SNAME()) " +
                   "EXEC " + QuoteIdentifier(databaseName) +
                   ".dbo.sp_changedbowner @loginame = N'" + EscapeStringLiteral(newDbOwner) + "', @map = false";
        }

        public void DetachDatabase()
        {
            using (var connection = new SqlConnection(_sqlInstanceConnString))
            {
                var command = new SqlCommand(
                    "EXEC master.dbo.sp_detach_db @dbname = @dbname, @keepfulltextindexfile = N'true'");
                command.Parameters.Add("@dbname", System.Data.SqlDbType.NVarChar, 128).Value = _databaseName;
                try
                {
                    connection.Open();
                    _lastSqlResult = true;
                }
                catch (Exception ex)
                {
                    _lastSqlException = ex;
                    _lastSqlResult = false;
                    return;
                }

                command.Connection = connection;
                command.ExecuteScalar();
                _lastSqlResult = true;
            }
        }

        public void CutAllConnections()
        {
            using (var connection = new SqlConnection(_sqlInstanceConnString))
            {
                var command = new SqlCommand(BuildCutAllConnectionsSql(_databaseName));
                try
                {
                    connection.Open();
                    _lastSqlResult = true;
                }
                catch (Exception ex)
                {
                    _lastSqlException = ex;
                    _lastSqlResult = false;
                    return;
                }

                command.Connection = connection;
                command.ExecuteScalar();
                _lastSqlResult = true;
            }
        }

        private static string BuildCutAllConnectionsSql(string databaseName)
        {
            return "ALTER DATABASE " + QuoteIdentifier(databaseName) +
                   " SET Single_User WITH ROLLBACK IMMEDIATE";
        }

        public void CreateDatabase(string dbName, string filenameOnSqlServer, int dbSizeInKb,
            int dbFileGrowthInKb, string dbLogname, string logFilenameOnSqlServer,
            int logSizeInKb, int logFileGrowthInPercent)
        {
            using (var connection = new SqlConnection(_sqlInstanceConnString))
            {
                var command = new SqlCommand(BuildCreateDatabaseSql(
                    dbName, filenameOnSqlServer, dbSizeInKb, dbFileGrowthInKb,
                    dbLogname, logFilenameOnSqlServer, logSizeInKb, logFileGrowthInPercent));
                try
                {
                    connection.Open();
                    _lastSqlResult = true;
                }
                catch (Exception ex)
                {
                    _lastSqlException = ex;
                    _lastSqlResult = false;
                    return;
                }

                command.Connection = connection;
                command.ExecuteScalar();
                _lastSqlResult = true;
            }
        }

        private static string BuildCreateDatabaseSql(
            string dbName, string filenameOnSqlServer, int dbSizeInKb, int dbFileGrowthInKb,
            string dbLogname, string logFilenameOnSqlServer, int logSizeInKb, int logFileGrowthInPercent)
        {
            return "CREATE DATABASE  " + QuoteIdentifier(dbName) + " ON  PRIMARY " +
                   "( NAME = N'" + EscapeStringLiteral(dbName) + "', FILENAME = N'" + EscapeStringLiteral(filenameOnSqlServer) +
                   "' , SIZE = " + dbSizeInKb + "KB , FILEGROWTH = " + dbFileGrowthInKb + "KB )" +
                   " LOG ON ( NAME = N'" + EscapeStringLiteral(dbLogname) + "', FILENAME = N'" + EscapeStringLiteral(logFilenameOnSqlServer) +
                   "' , SIZE = " + logSizeInKb + "KB , FILEGROWTH = " + logFileGrowthInPercent + "%)";
        }

        private static string QuoteIdentifier(string identifier)
        {
            if (string.IsNullOrEmpty(identifier))
                throw new ArgumentException("Identifier must not be null or empty.", nameof(identifier));

            foreach (char c in identifier)
            {
                if (!(char.IsLetterOrDigit(c) || c == '_' || c == '$' || c == '#' || c == '@'))
                    throw new ArgumentException(
                        "Identifier contains invalid characters: " + identifier, nameof(identifier));
            }

            return "[" + identifier.Replace("]", "]]") + "]";
        }

        private static string EscapeStringLiteral(string value)
        {
            if (value == null)
                return string.Empty;
            return value.Replace("'", "''");
        }

        /// <summary>
        /// Sendet ein T-Sql-Script zur aktuellen Datenbankinstanz (nicht zur Datenbank!)
        /// </summary>
        /// <param name="script">Zeichenkette, die das Skript enthält, das zur Datenbank gesendet werden soll.</param>
        public void SendSqlScript(string script)
        {
            using (var connection = new SqlConnection(_sqlInstanceConnString))
            {
                try
                {
                    connection.Open();
                    _lastSqlResult = true;
                }
                catch (Exception ex)
                {
                    _lastSqlException = ex;
                    _lastSqlResult = false;
                    return;
                }

                _lastSqlResult = true;
            }
        }

        public bool DatabaseExists => _databaseExists;
        public string FilenameOnSqlServer => _filenameOnSqlServer;
        public bool LastSqlResult => _lastSqlResult;
    }
}

File: src/MixNetFx/ADSqlSupportComponents/AdDatabaseManager.vb
Imports System.Data.SqlClient

Public Class ADSqlDatabaseManager

    Private mySqlInstanceConnString As String
    Private myDatabaseName As String
    Private myFilenameOnSqlServer As String
    Private myDatabaseExists As Boolean
    Private myLastSqlResult As Boolean
    Private myLastSqlException As Exception

    Sub New(ByVal SqlInstanceConnString As String, ByVal Databasename As String)
        mySqlInstanceConnString = SqlInstanceConnString
        myDatabaseName = Databasename
    End Sub

    Public Sub QueryProperties()
        Dim locConnection As New SqlConnection(mySqlInstanceConnString)
        Dim locCommand As SqlCommand

        Using locConnection
            Try
                locConnection.Open()
                myLastSqlResult = True
            Catch ex As Exception
                myLastSqlException = ex
                myLastSqlResult = False
                Return
            End Try

            locCommand = New SqlCommand("SELECT name FROM sys.databases where name='" & myDatabaseName & "'")
            locCommand.Connection = locConnection
            Dim locReader As SqlDataReader = locCommand.ExecuteReader()
            If locReader.HasRows Then
                myDatabaseExists = True
            Else
                myDatabaseExists = False
            End If
            locReader.Close()
        End Using

        If myDatabaseExists Then
            'Dateinamen ausfindig machen
            locConnection = New SqlConnection(mySqlInstanceConnString & "; Initial Catalog='" & myDatabaseName & "'")
            Using locConnection
                Try
                    locConnection.Open()
                    myLastSqlResult = True
                Catch ex As Exception
                    myLastSqlException = ex
                    myLastSqlResult = False
                    Return
                End Try

                locCommand = New SqlCommand("SELECT name, physical_name FROM sys.database_files where name='" & myDatabaseName & "'")
                locCommand.Connection = locConnection
                Dim locReader As SqlDataReader = locCommand.ExecuteReader()
                If locReader.HasRows Then
                    locReader.Read()
                    myFilenameOnSqlServer = locReader.GetString(locReader.GetOrdinal("physical_name"))
                End If
                locReader.Close()
            End Using
        End If
        myLastSqlResult = True
    End Sub

    Public Shared Function AttachDatabase(ByVal SqlInstanceConnString As String, ByVal Databasename As String, ByVal DatabaseFilename As String) As ADSqlDatabaseManager
        Return AttachDatabase(SqlInstanceConnString, Databasename, DatabaseFilename, Nothing, Nothing)
    End Function

    Public Shared Function AttachDatabase(ByVal SqlInstanceConnString As String, ByVal Databasename As String, ByVal DatabaseFilename As String, ByVal NewDbOwner As String) As ADSqlDatabaseManager
        Return AttachDatabase(SqlInstanceConnString, Databasename, DatabaseFilename, Nothing, NewDbOwner)
    End Function

    Public Shared Function AttachDatabase(ByVal SqlInstanceConnString As String, ByVal Databasename As String, _
                    ByVal DatabaseFilename As String, ByVal LogFilename As String, _
                    ByVal NewDbOwner As String) As ADSqlDatabaseManager

        Dim locConnection As New SqlConnection(SqlInstanceConnString)
        Dim locCommand As SqlCommand

        If String.IsNullOrEmpty(LogFilename) Then
            locCommand = New SqlCommand("CREATE DATABASE [" & Databasename & "] ON " & _
                "( FILENAME = N'" & DatabaseFilename & "' )" & _
                " FOR ATTACH_REBUILD_LOG", locConnection)
        Else
            locCommand = New SqlCommand("CREATE DATABASE [" & Databasename & "] ON " & _
                "( FILENAME = N'" & DatabaseFilename & "' )," & _
                "( FILENAME = N'" & LogFilename & "' )" & _
                " FOR ATTACH", locConnection)
        End If

        Using locConnection
            locConnection.Open()
            locCommand.ExecuteScalar()

            If Not String.IsNullOrEmpty(NewDbOwner) Then
                locCommand = New SqlCommand("if not exists (select name from master.sys.databases sd where name = N'" & Databasename & _
                "' and SUSER_SNAME(sd.owner_sid) = SUSER_SNAME() ) EXEC [" & Databasename & "].dbo.sp_changedbowner " & _
                "@loginame=N'" & NewDbOwner & "', @map=false", locConnection)
                locCommand.ExecuteScalar()
            End If
        End Using
        Return New ADSqlDatabaseManager(SqlInstanceConnString, Databasename)
    End Function

    Public Sub DetachDatabase()
        Dim locConnection As New SqlConnection(mySqlInstanceConnString)
        Dim locCommand As New SqlCommand("EXEC master.dbo.sp_detach_db @dbname = N'" & myDatabaseName & "', @keepfulltextindexfile=N'true'")

        Using locConnection
            Try
                locConnection.Open()
                myLastSqlResult = True
            Catch ex As Exception
                myLastSqlException = ex
                myLastSqlResult = False
                Return
            End Try

            locCommand.Connection = locConnection
            locCommand.ExecuteScalar()
            myLastSqlResult = True
        End Using
    End Sub

    Public Sub CutAllConnections()
        'Alle anderen Verbindungen sausen lassen - falls beispielsweise noch
        'der Visual Studio Server Explorer oder der Connection-Pool eine Verbindung auf sie hat.

        Dim locConnection As New SqlConnection(mySqlInstanceConnString)
        Dim locCommand As New SqlCommand("ALTER DATABASE " & myDatabaseName & " SET Single_User WITH ROLLBACK IMMEDIATE")

        Using locConnection
            Try
                locConnection.Open()
                myLastSqlResult = True
            Catch ex As Exception
                myLastSqlException = ex
                myLastSqlResult = False
                Return
            End Try

            locCommand.Connection = locConnection
            locCommand.ExecuteScalar()
            myLastSqlResult = True
        End Using
    End Sub

    Public Sub CreateDatabase(ByVal DbName As String, ByVal FilenameOnSqlServer As String, ByVal DbSizeInKb As Integer, _
                              ByVal DbFileGrowthInKb As Integer, ByVal DbLogname As String, ByVal LogFilenameOnSqlServer As String, _
                              ByVal LogSizeInKb As Integer, ByVal LogFileGrowthInPercent As Integer)

        Dim locConnection As New SqlConnection(mySqlInstanceConnString)
        Dim locCommand As New SqlCommand("CREATE DATABASE  [" & DbName & "] ON  PRIMARY " & _
                    "( NAME = N'" & DbName & "', FILENAME = N'" & FilenameOnSqlServer & _
                    "' , SIZE = " & DbSizeInKb & "KB , FILEGROWTH = " & DbFileGrowthInKb & "KB )" & _
                    " LOG ON ( NAME = N'" & DbLogname & "', FILENAME = N'" & LogFilenameOnSqlServer & _
                    "' , SIZE = " & LogSizeInKb & "KB , FILEGROWTH = " & LogFileGrowthInPercent & "%)")

        Using locConnection
            Try
                locConnection.Open()
                myLastSqlResult = True
            Catch ex As Exception
                myLastSqlException = ex
                myLastSqlResult = False
                Return
            End Try

            locCommand.Connection = locConnection
            locCommand.ExecuteScalar()
            myLastSqlResult = True
        End Using
    End Sub

    ''' <summary>
    ''' Sendet ein T-Sql-Script zur aktuellen Datenbankinstanz (nicht zur Datenbank!)
    ''' </summary>
    ''' <param name="Script">Zeichenkette, die das Skript enthält, das zur Datenbank gesendet werden soll.</param>
    ''' <remarks></remarks>
    Public Sub SendSqlScript(ByVal Script As String)

        Dim locConnection As New SqlConnection(mySqlInstanceConnString)

        Using locConnection
            Try
                locConnection.Open()
                myLastSqlResult = True
            Catch ex As Exception
                myLastSqlException = ex
                myLastSqlResult = False
                Return
            End Try

            myLastSqlResult = True
        End Using
    End Sub

    Public ReadOnly Property DatabaseExists() As Boolean
        Get
            Return myDatabaseExists
        End Get
    End Property

    Public ReadOnly Property FilenameOnSqlServer() As String
        Get
            Return myFilenameOnSqlServer
        End Get
    End Property

    Public ReadOnly Property LastSqlResult() As Boolean
        Get
            Return myLastSqlResult
        End Get
    End Property
End Class